File: backend/app/services/recruitment_service.py
import os
import logging
from uuid import UUID
from sqlalchemy import select, func, and_, delete
from sqlalchemy.orm import selectinload
from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import HTTPException, status

from app.models.recruitment import JobOffer, Candidate, Application, SavedJob
from app.models.scoring import CandidateScore, CVAnalysis
from app.models.interview import InterviewInvitation, InterviewSlot
from app.schemas.recruitment import JobOfferCreate, JobOfferUpdate, ApplicationUpdate

logger = logging.getLogger(__name__)


class RecruitmentService:

    async def create_job_offer(self, db: AsyncSession, data: JobOfferCreate, created_by: UUID) -> JobOffer:
        job = JobOffer(
            title=data.title,
            description=data.description,
            salary_min=data.salary_min,
            salary_max=data.salary_max,
            required_skills=data.required_skills,
            required_experience_years=data.required_experience_years,
            required_education_level=data.required_education_level,
            weight_skills=data.weight_skills,
            weight_experience=data.weight_experience,
            weight_education=data.weight_education,
            created_by=created_by
        )
        db.add(job)
        await db.flush()
        return job

    async def get_job_offers(self, db: AsyncSession, page: int = 1, size: int = 10, is_active: bool = True, search: str = None, sort_by: str = None) -> tuple[int, list[JobOffer]]:
        filters = [JobOffer.is_active == is_active]
        if search:
            filters.append(JobOffer.title.ilike(f"%{search}%") | JobOffer.description.ilike(f"%{search}%"))

        order = JobOffer.created_at.desc()
        if sort_by == 'oldest':
            order = JobOffer.created_at.asc()
        elif sort_by == 'title':
            order = JobOffer.title.asc()

        query = select(JobOffer).where(*filters).order_by(order)
        total = await db.scalar(select(func.count(JobOffer.id)).where(*filters))
        result = await db.execute(query.limit(size).offset((page - 1) * size))
        return total, result.scalars().all()

    async def get_job_offer(self, db: AsyncSession, job_id: UUID) -> JobOffer:
        result = await db.execute(select(JobOffer).where(JobOffer.id == job_id))
        job = result.scalar_one_or_none()
        if not job:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Job offer not found")
        return job

    async def update_job_offer(self, db: AsyncSession, job_id: UUID, data: JobOfferUpdate) -> JobOffer:
        job = await self.get_job_offer(db, job_id)
        if data.title is not None:
            job.title = data.title
        if data.description is not None:
            job.description = data.description
        if data.salary_min is not None:
            job.salary_min = data.salary_min
        if data.salary_max is not None:
            job.salary_max = data.salary_max
        if data.required_skills is not None:
            job.required_skills = data.required_skills
        if data.required_experience_years is not None:
            job.required_experience_years = data.required_experience_years
        if data.required_education_level is not None:
            job.required_education_level = data.required_education_level
        if data.weight_skills is not None:
            job.weight_skills = data.weight_skills
        if data.weight_experience is not None:
            job.weight_experience = data.weight_experience
        if data.weight_education is not None:
            job.weight_education = data.weight_education
        if data.is_active is not None:
            job.is_active = data.is_active
        await db.flush()
        return job

    async def delete_job_offer(self, db: AsyncSession, job_id: UUID) -> None:
        job = await self.get_job_offer(db, job_id)
        job.is_active = False
        await db.flush()

    async def get_or_create_candidate(self, db: AsyncSession, user_id: UUID, full_name: str, phone: str = None) -> Candidate:
        result = await db.execute(select(Candidate).where(Candidate.user_id == user_id))
        candidate = result.scalar_one_or_none()
        if not candidate:
            candidate = Candidate(user_id=user_id, full_name=full_name, phone=phone)
            db.add(candidate)
            await db.flush()
        return candidate

    async def apply_to_job(self, db: AsyncSession, candidate_id: UUID, job_offer_id: UUID, cv_file_path: str) -> Application:
        result = await db.execute(select(Application).where(
            and_(Application.candidate_id == candidate_id, Application.job_offer_id == job_offer_id)
        ))
        if result.scalar_one_or_none():
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Already applied")

        app = Application(candidate_id=candidate_id, job_offer_id=job_offer_id, cv_file_path=cv_file_path, status="PENDING")
        db.add(app)
        await db.flush()
        return app

    async def get_applications(self, db: AsyncSession, job_id: UUID = None, page: int = 1, size: int = 20) -> tuple[int, list[Application]]:
        query = select(Application).order_by(Application.created_at.desc())
        if job_id:
            query = query.where(Application.job_offer_id == job_id)

        count_query = select(func.count(Application.id))
        if job_id:
            count_query = count_query.where(Application.job_offer_id == job_id)
        total = await db.scalar(count_query)

        result = await db.execute(query.limit(size).offset((page - 1) * size))
        return total, result.scalars().all()

    async def get_my_applications(self, db: AsyncSession, user_id: UUID) -> list[dict]:
        """
        Return the current user's applications enriched with job offer details
        and optional score data. Returns dicts that map directly to MyApplicationResponse.
        """
        candidate_result = await db.execute(select(Candidate).where(Candidate.user_id == user_id))
        candidate = candidate_result.scalar_one_or_none()
        if not candidate:
            return []

        query = (
            select(Application, JobOffer, CandidateScore)
            .join(JobOffer, Application.job_offer_id == JobOffer.id)
            .outerjoin(CandidateScore, CandidateScore.application_id == Application.id)
            .where(Application.candidate_id == candidate.id)
            .order_by(Application.created_at.desc())
        )
        result = await db.execute(query)

        items: list[dict] = []
        for app, job, score in result.all():
            items.append({
                "id": app.id,
                "job_offer_id": app.job_offer_id,
                "candidate_id": app.candidate_id,
                "cv_file_path": app.cv_file_path,
                "status": app.status,
                "created_at": app.created_at,
                "updated_at": app.updated_at,
                "job_title": job.title,
                "job_description": job.description,
                "job_salary_min": job.salary_min,
                "job_salary_max": job.salary_max,
                "job_required_skills": job.required_skills,
                "job_required_experience_years": job.required_experience_years,
                "job_required_education_level": job.required_education_level,
                "job_is_active": job.is_active,
                "total_score": score.total_score if score else None,
                "skills_score": score.skills_score if score else None,
                "experience_score": score.experience_score if score else None,
                "education_score": score.education_score if score else None,
            })
        return items

    async def update_application_status(self, db: AsyncSession, app_id: UUID, data: ApplicationUpdate) -> Application:
        result = await db.execute(select(Application).where(Application.id == app_id))
        app = result.scalar_one_or_none()
        if not app:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application not found")
        app.status = data.status
        await db.flush()
        return app

    async def delete_application(self, db: AsyncSession, app_id: UUID) -> None:
        """
        Delete an application and clean up everything attached to it:
        the CV file on disk, the score, the CV analysis and any interview
        invitation (+ slots). Used by the admin "Supprimer candidature" action
        (CAND-08).
        """
        result = await db.execute(select(Application).where(Application.id == app_id))
        app = result.scalar_one_or_none()
        if not app:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Application not found")

        cv_path = app.cv_file_path

        # Free interview slots first (FK → invitations), then invitations.
        inv_ids = (
            await db.execute(
                select(InterviewInvitation.id).where(InterviewInvitation.application_id == app_id)
            )
        ).scalars().all()
        if inv_ids:
            await db.execute(delete(InterviewSlot).where(InterviewSlot.invitation_id.in_(inv_ids)))
            await db.execute(delete(InterviewInvitation).where(InterviewInvitation.id.in_(inv_ids)))

        # Remove derived scoring data (FK → applications, no DB cascade).
        await db.execute(delete(CandidateScore).where(CandidateScore.application_id == app_id))
        await db.execute(delete(CVAnalysis).where(CVAnalysis.application_id == app_id))

        await db.execute(delete(Application).where(Application.id == app_id))
        await db.flush()

        # Best-effort filesystem cleanup — never fail the request because of it.
        if cv_path:
            try:
                if os.path.exists(cv_path):
                    os.remove(cv_path)
            except OSError as exc:
                logger.warning(f"Could not delete CV file {cv_path}: {exc}")

    # ── Saved jobs (frontoffice bookmarks) ──────────────────────────────────

    async def get_saved_jobs(self, db: AsyncSession, user_id: UUID) -> list[JobOffer]:
        result = await db.execute(
            select(JobOffer)
            .join(SavedJob, SavedJob.job_offer_id == JobOffer.id)
            .where(SavedJob.user_id == user_id)
            .order_by(SavedJob.created_at.desc())
        )
        return list(result.scalars().all())

    async def save_job(self, db: AsyncSession, user_id: UUID, job_id: UUID) -> JobOffer:
        job = await self.get_job_offer(db, job_id)  # 404 if missing
        existing = await db.execute(
            select(SavedJob).where(
                and_(SavedJob.user_id == user_id, SavedJob.job_offer_id == job_id)
            )
        )
        if existing.scalar_one_or_none() is None:
            db.add(SavedJob(user_id=user_id, job_offer_id=job_id))
            await db.flush()
        return job

    async def unsave_job(self, db: AsyncSession, user_id: UUID, job_id: UUID) -> None:
        await db.execute(
            delete(SavedJob).where(
                and_(SavedJob.user_id == user_id, SavedJob.job_offer_id == job_id)
            )
        )
        await db.flush()


recruitment_service = RecruitmentService()

File: frontend/src/app/core/services/recruitment.service.ts
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiService } from './api.service';
import { environment } from '../../../environments/environment';

export interface JobOffer {
  id: string;
  title: string;
  description: string;
  salary_min: number | null;
  salary_max: number | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  required_skills?: string[];
  required_experience_years?: number;
  required_education_level?: string;
  weight_skills?: number;
  weight_experience?: number;
  weight_education?: number;
}

export interface Application {
  id: string;
  candidate_id: string;
  candidate_name?: string;
  job_offer_id: string;
  job_offer_title?: string;
  cv_file_path: string;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface MyApplication {
  id: string;
  job_offer_id: string;
  candidate_id: string;
  cv_file_path: string;
  status: string;
  created_at: string;
  updated_at: string;
  job_title: string;
  job_description: string;
  job_salary_min: number | null;
  job_salary_max: number | null;
  job_required_skills: string[] | null;
  job_required_experience_years: number | null;
  job_required_education_level: string | null;
  job_is_active: boolean;
  total_score: number | null;
  skills_score: number | null;
  experience_score: number | null;
  education_score: number | null;
}

export interface JobOfferWithCandidateCount extends JobOffer {
  candidate_count: number;
}

export interface PaginatedJobs {
  total: number;
  page: number;
  size: number;
  items: JobOffer[];
}

export interface CandidateRanking {
  candidate_id: string;
  candidate_name?: string;
  application_id: string;
  total_score: number;
  rank: number | null;
  skills_score: number;
  experience_score: number;
  education_score: number;
  score_details: ScoreDetails | null;
  // Set when the backend finds an active interview invitation for this app.
  interview_status?: 'PENDING' | 'CONFIRMED' | null;
  interview_invitation_id?: string | null;
  interview_confirmed_at?: string | null;
}

export interface RankingResponse {
  job_offer_id: string;
  total_candidates: number;
  ranking: CandidateRanking[];
}

export interface ScoreDetails {
  matched_skills: string[];
  missing_skills: string[];
  required_skills: string[];
  extracted_years: number | null;
  required_years: number | null;
  extracted_education: string | null;
  required_education: string | null;
  weights: { skills: number; experience: number; education: number };
  strengths: string[];
  weaknesses: string[];
  recommendation: string;
}

export interface CandidateScoreResponse {
  id: string;
  application_id: string;
  job_offer_id: string;
  candidate_id: string;
  total_score: number;
  skills_score: number;
  experience_score: number;
  education_score: number;
  rank: number | null;
  score_details: ScoreDetails | null;
  computed_at: string;
}

export interface CVAnalysisResponse {
  id: string;
  application_id: string;
  candidate_id: string;
  extracted_skills: string[];
  extracted_experience_years: number | null;
  extracted_education_level: string;
  extracted_job_titles: string[];
  extracted_keywords: string[];
  is_parsed: boolean;
  parsed_at: string | null;
  created_at: string;
}

export interface AnalysisStartedResponse {
  message: string;
  application_id: string | null;
  job_id: string | null;
}

export interface LabelCount { label: string; count: number; }
export interface ApplicationsTimeline {
  series: { date: string; label: string; count: number }[];
  total: number;
}
export interface CandidateAnalytics {
  experience_buckets: LabelCount[];
  score_bands: LabelCount[];
  education: LabelCount[];
  top_skills: LabelCount[];
  analyzed_cvs: number;
  scored_candidates: number;
  generated_at: string;
}

export interface RecruitmentSummary {
  total_jobs: number;
  active_jobs: number;
  total_applications: number;
  applications_by_status: Record<string, number>;
  acceptance_rate: number;
  average_score: number | null;
  top_jobs: { job_offer_id: string; title: string; application_count: number }[];
  generated_at: string;
}

// ─── Interview workflow ───
export type InvitationStatus = 'PENDING' | 'CONFIRMED' | 'EXPIRED' | 'CANCELLED';
export type SlotState = 'AVAILABLE' | 'PROPOSED' | 'RESERVED';

export interface InterviewSlot {
  id: string;
  invitation_id: string;
  start_at: string;
  end_at: string;
  is_selected: boolean;
}

export interface InvitationResponse {
  id: string;
  application_id: string;
  token: string;
  status: InvitationStatus;
  message: string | null;
  expires_at: string;
  confirmed_slot_id: string | null;
  confirmed_at: string | null;
  cancelled_at: string | null;
  cancellation_reason: string | null;
  created_at: string;
  updated_at: string;
  slots: InterviewSlot[];
  candidate_name: string | null;
  candidate_email: string | null;
  job_title: string | null;
  public_url: string;
}

export interface PublicInvitationView {
  status: InvitationStatus;
  expires_at: string;
  message: string | null;
  candidate_name: string;
  job_title: string;
  job_description: string | null;
  company_name: string;
  slots: InterviewSlot[];
  confirmed_slot: InterviewSlot | null;
}

export interface CalendarSlot {
  id: string;
  invitation_id: string;
  application_id: string;
  start_at: string;
  end_at: string;
  state: SlotState;
  candidate_name: string | null;
  job_title: string | null;
  invitation_status: InvitationStatus;
  confirmed_at: string | null;
}

export interface CalendarResponse {
  total: number;
  items: CalendarSlot[];
}

@Injectable({ providedIn: 'root' })
export class RecruitmentService {
  constructor(private api: ApiService) {}

  getJobs(page = 1, size = 10, search?: string, sortBy?: string): Observable<PaginatedJobs> {
    const params: Record<string, string | number> = { page, size };
    if (search) params['search'] = search;
    if (sortBy) params['sort_by'] = sortBy;
    return this.api.get<PaginatedJobs>('recruitment/jobs', params);
  }

  getJob(jobId: string): Observable<JobOffer> {
    return this.api.get<JobOffer>(`recruitment/jobs/${jobId}`);
  }

  createJob(data: any): Observable<JobOffer> {
    return this.api.post<JobOffer>('recruitment/jobs', data);
  }

  updateJob(jobId: string, data: any): Observable<JobOffer> {
    return this.api.patch<JobOffer>(`recruitment/jobs/${jobId}`, data);
  }

  deleteJob(jobId: string): Observable<void> {
    return this.api.delete<void>(`recruitment/jobs/${jobId}`);
  }

  applyToJob(jobId: string, cvFile: File): Observable<Application> {
    const formData = new FormData();
    formData.append('cv_file', cvFile);
    return this.api.postFormData<Application>(`recruitment/apply/${jobId}`, formData);
  }

  getApplications(jobId?: string, page = 1, size = 20): Observable<Application[]> {
    const params: Record<string, string | number> = { page, size };
    if (jobId) params['job_id'] = jobId;
    return this.api.get<Application[]>('recruitment/applications', params);
  }

  /** Authenticated user's own applications, enriched with job offer details + score. */
  getMyApplications(): Observable<MyApplication[]> {
    return this.api.get<MyApplication[]>('recruitment/my-applications');
  }

  getApplicationsByJob(jobId: string): Observable<Application[]> {
    return this.api.get<Application[]>(`recruitment/jobs/${jobId}/applications`);
  }

  updateApplication(appId: string, status: string): Observable<Application> {
    return this.api.patch<Application>(`recruitment/applications/${appId}`, { status });
  }

  /** CAND-08 — hard-delete an application (Admin only). */
  deleteApplication(appId: string): Observable<void> {
    return this.api.delete<void>(`recruitment/applications/${appId}`);
  }

  /** Full update of a job offer (PUT). */
  replaceJob(jobId: string, data: any): Observable<JobOffer> {
    return this.api.put<JobOffer>(`recruitment/jobs/${jobId}`, data);
  }

  // ─── Saved jobs (frontoffice bookmarks) ───
  getSavedJobs(): Observable<JobOffer[]> {
    return this.api.get<JobOffer[]>('recruitment/saved-jobs');
  }

  saveJob(jobId: string): Observable<JobOffer> {
    return this.api.post<JobOffer>(`recruitment/saved-jobs/${jobId}`, {});
  }

  unsaveJob(jobId: string): Observable<void> {
    return this.api.delete<void>(`recruitment/saved-jobs/${jobId}`);
  }

  /** Export applications as a CSV blob (optionally filtered by job). */
  exportApplications(jobId?: string): Observable<Blob> {
    const qs = jobId ? `?job_id=${jobId}` : '';
    return this.api.http.get(
      `${environment.apiUrl}/recruitment/applications/export${qs}`,
      { responseType: 'blob' },
    );
  }

  getCVDownloadUrl(cvPath: string): string {
    const encodedPath = encodeURIComponent(cvPath);
    return `http://localhost:8000/api/v1/recruitment/cv-download/${encodedPath}`;
  }

  getCVPreviewUrl(cvPath: string): string {
    const encodedPath = encodeURIComponent(cvPath);
    return `http://localhost:8000/api/v1/recruitment/cv-preview/${encodedPath}`;
  }

  downloadCVAsBlob(cvPath: string): Observable<Blob> {
    const encodedPath = encodeURIComponent(cvPath);
    return this.api.http.get(`http://localhost:8000/api/v1/recruitment/cv-preview/${encodedPath}`,
      { responseType: 'blob' });
  }

  analyzeApplication(appId: string): Observable<AnalysisStartedResponse> {
    return this.api.post<AnalysisStartedResponse>(`recruitment/applications/${appId}/analyze`, {});
  }

  /** Trigger CV analysis + scoring for all applications of a job. */
  analyzeAllForJob(jobId: string): Observable<AnalysisStartedResponse> {
    return this.api.post<AnalysisStartedResponse>(`recruitment/jobs/${jobId}/analyze-all`, {});
  }

  /** Live recruitment KPIs (reports module). */
  getRecruitmentSummary(): Observable<RecruitmentSummary> {
    return this.api.get<RecruitmentSummary>('reports/recruitment-summary');
  }

  /** Candidate distributions (experience, AI score band, education, top skills). */
  getCandidateAnalytics(): Observable<CandidateAnalytics> {
    return this.api.get<CandidateAnalytics>('reports/candidate-analytics');
  }

  /** Daily applications over the last N days (evolution chart). */
  getApplicationsTimeline(days = 14): Observable<ApplicationsTimeline> {
    return this.api.get<ApplicationsTimeline>('reports/applications-timeline', { days });
  }

  scheduleInterview(
    appId: string,
    slots: string[],
    message: string,
    durationMinutes = 30,
    expiresInHours = 48,
  ): Observable<{ message: string }> {
    return this.api.post(`recruitment/applications/${appId}/schedule-interview`, {
      slots,
      message,
      duration_minutes: durationMinutes,
      expires_in_hours: expiresInHours,
    });
  }

  // ─── Public (no auth) — invitation confirm page ───
  getPublicInvitation(token: string): Observable<PublicInvitationView> {
    return this.api.get<PublicInvitationView>(`interview/confirm/${token}`);
  }

  confirmInvitationSlot(token: string, slotId: string): Observable<PublicInvitationView> {
    return this.api.post<PublicInvitationView>(`interview/confirm/${token}`, { slot_id: slotId });
  }

  /** Lightweight check: does this application already have a PENDING/CONFIRMED invitation? */
  getActiveInvitationForApplication(appId: string): Observable<{
    active: boolean;
    id?: string;
    status?: InvitationStatus;
    token?: string;
    expires_at?: string;
    confirmed_slot_id?: string | null;
    confirmed_at?: string | null;
    public_url?: string;
  }> {
    return this.api.get(`interview/applications/${appId}/active-invitation`);
  }

  // ─── RH ───
  listInvitations(filter?: { status?: InvitationStatus; applicationId?: string }): Observable<InvitationResponse[]> {
    const params: Record<string, string | number> = {};
    if (filter?.status)        params['status']         = filter.status;
    if (filter?.applicationId) params['application_id'] = filter.applicationId;
    return this.api.get<InvitationResponse[]>('interview/invitations', params);
  }

  getInvitation(invitationId: string): Observable<InvitationResponse> {
    return this.api.get<InvitationResponse>(`interview/invitations/${invitationId}`);
  }

  cancelInvitation(invitationId: string, reason: string = ''): Observable<InvitationResponse> {
    return this.api.post<InvitationResponse>(`interview/invitations/${invitationId}/cancel`, { reason });
  }

  getInterviewCalendar(filter?: { dateFrom?: string; dateTo?: string; state?: SlotState }): Observable<CalendarResponse> {
    const params: Record<string, string> = {};
    if (filter?.dateFrom) params['date_from'] = filter.dateFrom;
    if (filter?.dateTo)   params['date_to']   = filter.dateTo;
    if (filter?.state)    params['state']     = filter.state;
    return this.api.get<CalendarResponse>('interview/calendar', params);
  }

  /** État de l'intégration Google Calendar côté serveur. */
  getGoogleCalendarStatus(): Observable<{ configured: boolean; calendar_id: string | null; reason: string | null }> {
    return this.api.get('interview/google-status');
  }

  /** (Re)pousse un entretien confirmé vers Google Calendar. */
  syncInvitationToGoogle(invitationId: string): Observable<{ synced: boolean; google_event_id: string; already: boolean }> {
    return this.api.post(`interview/invitations/${invitationId}/sync-google`, {});
  }

  /** Télécharge le fichier .ics (RFC 5545) d'un entretien confirmé. */
  downloadInvitationIcs(invitationId: string): Observable<Blob> {
    return this.api.http.get(`${environment.apiUrl}/interview/invitations/${invitationId}/ics`, {
      responseType: 'blob',
    });
  }

  rejectApplication(appId: string, message: string = ''): Observable<{ message: string }> {
    return this.api.post(`recruitment/applications/${appId}/reject`, { message });
  }

  startNegotiation(appId: string, employerOffer: number): Observable<any> {
    return this.api.post(`recruitment/applications/${appId}/start-negotiation`, { employer_offer: employerOffer });
  }

  getRankedCandidates(jobId: string): Observable<RankingResponse> {
    return this.api.get<RankingResponse>(`recruitment/jobs/${jobId}/ranking`);
  }

  getScoreBreakdown(appId: string): Observable<CandidateScoreResponse> {
    return this.api.get<CandidateScoreResponse>(`recruitment/applications/${appId}/score`);
  }

  getExtractedCVData(appId: string): Observable<CVAnalysisResponse> {
    return this.api.get<CVAnalysisResponse>(`recruitment/applications/${appId}/analysis`);
  }
}
